L1Trigger/L1THGCal/interface/be_algorithms/HGCalTowerMap2DImpl.h

```diff
@@ -0,0 +1,53 @@
+#ifndef __L1Trigger_L1THGCal_HGCalTowerMap2DImpl_h__
+#define __L1Trigger_L1THGCal_HGCalTowerMap2DImpl_h__
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+
+#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
+#include "DataFormats/L1THGCal/interface/HGCalTowerMap.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerTools.h"
+
+
+class HGCalTowerMap2DImpl{
+
+ public:
+
+  HGCalTowerMap2DImpl( const edm::ParameterSet &conf); 
+  
+  void resetTowerMaps( );
+
+  void buildTowerMap2D( const std::vector<edm::Ptr<l1t::HGCalTriggerCell>> & triggerCellsPtrs,
+			l1t::HGCalTowerMapBxCollection & towermaps
+			);
+
+
+  void eventSetup(const edm::EventSetup& es) 
+    {
+        triggerTools_.setEventSetup(es);
+    }
+
+  
+ private:
+
+  static const int kLayersEE_ = 28;
+  static const int kLayersFH_ = 12;  
+  static const unsigned kLayersBH_ = 12;
+  static const unsigned kLayers_ = kLayersEE_+kLayersFH_+kLayersBH_;
```

Comment:
Could add this information in `triggerTools` (already `kLayersEE` and `kLayersFH` are available there)

Reply:
The way this information is recovered in triggerTools is with
`(geom_->ee/fhTopology().dddConstants()).layers(true)`
Unfortunately it does not work out of the box with BH because it uses `HcalTopology` instead of `HGCalTopology`.
Has this issue already been solved elsewhere?

Reply:
This is correct, you can find an example for instance here:
https://github.com/PFCal-dev/cmssw/blob/hgc-tpg-devel-CMSSW_9_4_0_pre2/RecoLocalCalo/HGCalRecAlgos/src/RecHitTools.cc#L131